src/librustc/ty/layout.rs

```diff
@@ -542,567 +597,222 @@ impl Primitive {
-        // Field 5 would be the first element, so memory_index is i:
-        // Note: if we didn't optimize, it's already right.
+    /// Struct-like placement, with precomputed offsets.
+    Arbitrary {
+        /// Offsets for the first byte of each field,
+        /// ordered to match the source definition order.
+        /// This vector does not go in increasing order.
+        // FIXME(eddyb) use small vector optimization for the common case.
+        offsets: Vec<Size>,
 
-        if optimize {
-            ret.memory_index = vec![0; inverse_memory_index.len()];
+        /// Maps source order field indices to memory order indices,
+        /// depending how fields were permuted.
+        // FIXME(camlorn) also consider small vector  optimization here.
+        memory_index: Vec<u32>
+    }
+}
 
-            for i in 0..inverse_memory_index.len() {
-                ret.memory_index[inverse_memory_index[i] as usize]  = i as u32;
+impl FieldPlacement {
+    pub fn count(&self) -> usize {
+        match *self {
+            FieldPlacement::Union(count) => count,
+            FieldPlacement::Array { count, .. } => {
+                let usize_count = count as usize;
+                assert_eq!(usize_count as u64, count);
+                usize_count
             }
-        } else {
-            ret.memory_index = inverse_memory_index;
+            FieldPlacement::Arbitrary { ref offsets, .. } => offsets.len()
         }
-
-        Ok(ret)
     }
 
-    /// Get the size with trailing alignment padding.
-    pub fn stride(&self) -> Size {
-        self.min_size.abi_align(self.align)
+    pub fn offset(&self, i: usize) -> Size {
+        match *self {
+            FieldPlacement::Union(_) => Size::from_bytes(0),
+            FieldPlacement::Array { stride, count } => {
+                let i = i as u64;
+                assert!(i < count);
+                stride * i
+            }
+            FieldPlacement::Arbitrary { ref offsets, .. } => offsets[i]
+        }
     }
 
-    /// Determine whether a structure would be zero-sized, given its fields.
-    fn would_be_zero_sized<I>(dl: &TargetDataLayout, fields: I)
-                              -> Result<bool, LayoutError<'tcx>>
-    where I: Iterator<Item=Result<&'a Layout, LayoutError<'tcx>>> {
-        for field in fields {
-            let field = field?;
-            if field.is_unsized() || field.size(dl).bytes() > 0 {
-                return Ok(false);
+    pub fn memory_index(&self, i: usize) -> usize {
+        match *self {
+            FieldPlacement::Union(_) |
+            FieldPlacement::Array { .. } => i,
+            FieldPlacement::Arbitrary { ref memory_index, .. } => {
+                let r = memory_index[i];
+                assert_eq!(r as usize as u32, r);
+                r as usize
             }
         }
-        Ok(true)
     }
 
-    /// Get indices of the tys that made this struct by increasing offset.
+    /// Get indices of the fields by increasing offset.
```

Comment:
comment nit: Get memory indices of the fields by increasing offset.

Reply:
They're source indices, in memory offset order, so I don't see how "memory indices" works there.